fractal_tasks_core/tasks/cellpose_utils.py

```diff
@@ -0,0 +1,468 @@
+            augmentation to tile images with overlap.
+        net_avg: Parameter of `CellposeModel` class. Whether to use cellpose
+            net averaging to run the 4 built-in networks (useful for `nuclei`,
+            `cyto` and `cyto2`, not sure it works for the others).
+        use_gpu: If `False`, always use the CPU; if `True`, use the GPU if
+            possible (as defined in `cellpose.core.use_gpu()`) and fall-back
+            to the CPU otherwise.
+        batch_size: number of 224x224 patches to run simultaneously on the GPU
+            (can make smaller or bigger depending on GPU memory usage)
+        invert: invert image pixel intensity before running network (if True,
+            image is also normalized)
+        tile: tiles image to ensure GPU/CPU memory usage limited (recommended)
+        tile_overlap: fraction of overlap of tiles when computing flows
+        resample: run dynamics at original image size (will be slower but
+            create more accurate boundaries)
+        interp: interpolate during 2D dynamics (not available in 3D)
+            (in previous versions it was False, now it defaults to True)
+        stitch_threshold: if stitch_threshold>0.0 and not do_3D and equal
+            image sizes, masks are stitched in 3D to return volume segmentation
+    """
+
+    cellprob_threshold: float = 0.0
+    flow_threshold: float = 0.4
+    anisotropy: Optional[float] = None
+    min_size: int = 15
+    augment: bool = False
+    net_avg: bool = False
+    use_gpu: bool = True
+    batch_size: int = 8
+    invert: bool = False
+    tile: bool = True
+    tile_overlap: float = 0.1
+    resample: bool = True
+    interp: bool = True
+    stitch_threshold: float = 0.0
+
+
+class CellposeChannel1InputModel(ChannelInputModel):
+    """
+    Channel input for cellpose with normalization options.
+
+    Attributes:
+        wavelength_id: Unique ID for the channel wavelength, e.g. `A01_C01`.
+            Can only be specified if label is not set.
+        label: Name of the channel. Can only be specified if wavelength_id is
+            not set.
+        normalize: Validator to handle different normalization scenarios for
+            Cellpose models
+    """
+
+    normalize: CellposeCustomNormalizer = Field(
+        default_factory=CellposeCustomNormalizer
+    )
+
+    def get_omero_channel(self, zarr_url) -> OmeroChannel:
+        try:
+            return get_channel_from_image_zarr(
+                image_zarr_path=zarr_url,
+                wavelength_id=self.wavelength_id,
+                label=self.label,
+            )
+        except ChannelNotFoundError as e:
+            logger.warning(
+                f"Channel with wavelength_id: {self.wavelength_id} "
+                f"and label: {self.label} not found, exit from the task.\n"
+                f"Original error: {str(e)}"
+            )
+            return None
+
+
+class CellposeChannel2InputModel(BaseModel):
```

Comment:
I may be missing something obvious here: how does `CellposeChannel2InputModel` differ from `CellposeChannel1InputModel`?

1. They both have optional `wavelength_id` and `label` attributes.
2. They both have a validator for this two (mutually-exclusive) attributes
3. They both have a `get_omero_channel`.
4. Their `normalize` attribute is the same.

The differences I find are
1. A different error message in `get_omero_channel` (is it important?)
2. The `is_set` method of the second model.

I suggest one of the following:
* Ideally, we have a single `CellposeChannelInputModel` model. The fact that `is_set` is only used when it refers to the second channel is not an issue. We would loose the different error messages though.
* If we need to keep different error messages, let's make `CellposeChannelInput2Model` inherit from either `ChannelInputModel` or `CellposeChannelInput1Model` - and then we would only redefine `get_omero_channel` and add the `is_set` method.


I suggest to merge the two models

Reply:
The core difference is that Channel 1 is required and Channel 2 is optional. Therefore, Channel1 is non optional input, but channel 2 has only either optional string arguments (wavelength id or label) or objects with a default. Channel 2's validation also doesn't enforce that it's specified (and thus has the new `is_set` method).

We could do it via the `Optional` input for channel 2, but we tried to avoid this (because the complex Channel option is tricky to handle in Fractal web & the json manifest afterwards).

-----

If we can achieve a required Channel 1 & an optional Channel 2 with a single model though, that's even better!

Reply:
OK, thanks!

> channel 2 has only either optional string arguments (wavelength id or label) or objects with a default

This is true for both channels 1 and 2 - as far as I understand.

> Channel 2's validation also doesn't enforce that it's specified (and thus has the new is_set method).

My bad, I missed this one (I saw the same `mutually_exclusive_channel_attributes` method name and thought they were the same for the two models). Now I see.

> We could do it via the Optional input for channel 2, but we tried to avoid this (because the complex Channel option is tricky to handle in Fractal web & the json manifest afterwards).
> If we can achieve a required Channel 1 & an optional Channel 2 with a single model though, that's even better!

I don't have anything trivial to suggest, but I am having a look now

Reply:
(this is now discussed in the main comments)